### PR TITLE
Fix auto restart bugs

### LINKED ISSUE #4945: Log count number doesn't increment for auto restart?
I think the log count number might not be 100% airtight, got this situation while running `tests/functional/restart/34-auto-restart-basic.t`:
```
01-start-01.log
02-start-01.log
02-restart-02.log
03-restart-02.log
```
(sorted by creation time ascending)

When I re-ran it passed, but this shouldn't be flaky

_Originally posted by @MetRonnie in https://github.com/cylc/cylc-flow/issues/4836#issuecomment-1156767719_

### PR DESCRIPTION
**Pull request by MetRonnie**
- Fix bug where sometimes the scheduler log would rollover unnecessarily on auto restart, with clashing log count number in the filename (closes #4945)
- Ensure the workflow restarts with the same verbosity.
- Fix bug where the workflow was failing to abort if auto restart was attempted in no-detach mode
- Prevent unexpected exceptions raised during auto restart from being swallowed

Reviewers should run `tests/functional/restart` tests

**Check List**

- [x] I have read `CONTRIBUTING.md` and added my name as a Code Contributor.
- [x] Contains logically grouped changes (else tidy your branch by rebase).
- [x] Does not contain off-topic changes (use other PRs for other changes).
- [x] Applied any dependency changes to both `setup.cfg` and `conda-environment.yml`.
- [x] Tests are included (or explain why tests are not needed).
- [x] `CHANGES.md` entry included if this is a change that can affect users
- [x] [Cylc-Doc](https://github.com/cylc/cylc-doc) pull request opened if required at cylc/cylc-doc/pull/XXXX.
- [ ] If this is a bug fix, PRs raised to both master and the relevant maintenance branch.

### CODE REVIEW

**Review by datamel — COMMENTED** [review]
Can I check @MetRonnie... on this branch when running the restart tests I'm still getting failures on:
``` 
tests/functional/restart/41-auto-restart-local-jobs.t
tests/functional/restart/42-auto-restart-ping-pong.t
tests/functional/restart/37-auto-restart-delay.t
tests/functional/restart/34-auto-restart-basic.t
```
Are these passing for you?
I'm still getting the log number 02 repeating for both start and restart.

File: cylc/flow/scheduler.py [hunk 1]
```diff
@@ -1393,17 +1393,16 @@ async def workflow_shutdown(self):
...
-                        and itask.summary['job_runner_name']
-                        and not is_remote_platform(itask.platform)
-                        and self.task_job_mgr.job_runner_mgr
-                        .is_job_local_to_host(
-                            itask.summary['job_runner_name'])
+                    itask.state(TASK_STATUS_PREPARING, *TASK_STATUSES_ACTIVE)
+                    and itask.summary['job_runner_name']
+                    and not is_remote_platform(itask.platform)
+                    and self.task_job_mgr.job_runner_mgr.is_job_local_to_host(
+                        itask.summary['job_runner_name'])
```
[the remark this answers is not in the document]
Comment (MetRonnie) [reply cylc/flow/scheduler.py]: (I've removed that attempted fix from this PR and working on it on new branch)

**Review by oliver-sanders — APPROVED** [review]
LGTM.

* Tested by bouncing a workflow between two servers with the log rollover size turned down, restart and rollover numbers behaved correctly.
* Ran the relevant functional/flakyfunctional tests locally (they don't run on CI).

**Review by datamel — APPROVED** [review]
Thanks @MetRonnie. Great changes, now working well for me.

**Pull request merged**